src/hotspot/share/prims/jvmtiThreadState.cpp

```diff
@@ -358,11 +358,13 @@ JvmtiVTMSTransitionDisabler::VTMS_transition_disable_for_all() {
 
     // Block while some mount/unmount transitions are in progress.
     // Debug version fails and prints diagnostic information.
-    while (_VTMS_transition_count > 0) {
-      if (ml.wait(10)) {
-        attempts--;
+    for (JavaThreadIteratorWithHandle jtiwh; JavaThread *jt = jtiwh.next(); ) {
+      while (jt->VTMS_transition_mark()) {
+        if (ml.wait(10)) {
+          attempts--;
+        }
+        DEBUG_ONLY(if (attempts == 0) break;)
```

Comment:
Thank you for the comment. It is thinking in a right direction.
Each `JavaThread` set the `VTM_transition_mark` only once and then checks for disable counters:
 - `_VTMS_transition_disable_for_all_count`
 - `java_lang_Thread::VTMS_transition_disable_count(vth())`
 
If any of the disable counters is not zero then each `JavaThread` clears the optimistically set mark and continues under protection of the `JvmtiVTMSTransition_lock`.

Reply:
I'm not sure this answered Chris' query properly. Or I'm reading Chris' query wrong.

Perhaps this is not what Chris had in mind, but I'm wondering what happens in some
Thread-A when it is checked and passed by but then Thread-A sets the flag in itself
after the for-loop has passed it by. Does that Thread-A flag value get lost?

Reply:
> Perhaps this is not what Chris had in mind, but I'm wondering what happens in some
> Thread-A when it is checked and passed by but then Thread-A sets the flag in itself
> after the for-loop has passed it by. Does that Thread-A flag value get lost?

Thank you for the question.
The Thread-A sets the flag optimistically and then re-checks if `sync_protocol_enabled()` and any disabler exists. It can be global disbaler (`_VTMS_transition_disable_for_all_count > 0`) or disabler of `Thread-A` only (`java_lang_Thread::VTMS_transition_disable_count(vth()) > 0`). If any disabler exists then `Thread-A` clears the optimistic settings and goes with the pessimistic approach under protection of `JvmtiVTMSTransition_lock`.

Please, let me know if you still have questions.

Reply:
This algorithm sounds correct. Thanks for closing the loop on my belated comment.